Basic: record layering violation for linking

compnerd opened the pull request:
When linking on Linux, we would fail with unresolved symbol references
to  `swift::FrontendStatsTracer::getTraceFormatter<T>`.  The use of the
types, which are defined in swiftSIL occur in swiftBasic.  To avoid the
circular dependency, introduce interface link dependencies.  This allows
the consumers of swiftBasic to also link against swiftSIL at the right
time, resolving the symbol.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd commented:
CC: @gottesmm @jrose-apple

compnerd commented:
@swift-ci please test

gottesmm commented:
@compnerd why doesn't this happen on Darwin?

jrose-apple reviewed (dismissed):
Basic is the bottom of the library hierarchy; this really is a circular dependency. We'll have to find another answer.

jrose-apple commented:
@gottesmm: ld64 doesn't pay attention to dependencies between static libraries, so none of this happens on Darwin. If we built dylibs we'd see all of these problems.

(This is hardly the only circularity but it's a particularly egregious one because it involves Basic.)

graydon reviewed (changes requested):
Yeah no this isn't right. The slightly subtle arrangement of declaring swift::FrontendStatsTracer::getTraceFormatter<T> instances in Basic, against forward-declared higher-level types, and then only defining those in the higher-layer libraries, is specifically done to avoid linking errors and circular dependency. If that is somehow failing, I made a mistake or something. Can you post the exact error you're getting? (I develop on 50/50 linux and darwin and this is not happening on my end).

gottesmm commented:
@jrose-apple that wasn't clear to me. It could have also been due to ld64 not iterating. I guess I should have been more specific with my question.

gottesmm commented:
*erm ld64 iterating and gold not doing so in the case of linking in a dylib. I didn't see it was an issue with static libraries. This reminds me of crazy ideas I have had in the past for identifying layering violations via something in lld. But that is an idea for another time.

compnerd commented:
@graydon it just gives a bunch of unresolved symbol errors when building with gold.  The next time I do a full build, Ill try to grab the error message.

AndrewSB commented:
Here's the error:

```
--- Building tests for swift ---
+ /usr/bin/cmake --build /home/asb/Developer/swift/build/Ninja-DebugAssert/swift-linux-x86_64 -- -j24 SwiftUnitTests
[4/9] Linking CXX executable unittests/Driver/SwiftDriverTests
FAILED: unittests/Driver/SwiftDriverTests
: && /usr/bin/clang++   -Wno-unknown-warning-option -Werror=unguarded-availability-new -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++11 -Wall -W -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wcovered-switch-default -Wdelete-non-virtual-dtor -Wstring-conversion -fcolor-diagnostics -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -g  -latomic -fuse-ld=lld unittests/Driver/CMakeFiles/SwiftDriverTests.dir/DependencyGraphTests.cpp.o  -o unittests/Driver/SwiftDriverTests -L/home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/./lib -Wl,-rpath,/home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/./lib /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMSupport.a -lpthread /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libgtest_main.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libgtest.a -lpthread lib/libswiftDriver.a -lpthread lib/libswiftFrontend.a lib/libswiftMigrator.a lib/libswiftIDE.a lib/libswiftFrontend.a lib/libswiftMigrator.a lib/libswiftIDE.a lib/libswiftParseSIL.a lib/libswiftSIL.a lib/libswiftSema.a lib/libswiftSerialization.a lib/libswiftClangImporter.a lib/libswiftParse.a lib/libswiftIndex.a lib/libswiftAST.a lib/libswiftMarkup.a /home/asb/Developer/swift/build/Ninja-DebugAssert/cmark-linux-x86_64/src/libcmark.a lib/libswiftSyntax.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMDebugInfoDWARF.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMX86CodeGen.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMX86AsmParser.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMX86Desc.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMX86AsmPrinter.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMX86Disassembler.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMX86Info.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMX86Utils.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMARMCodeGen.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMARMAsmParser.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMARMDisassembler.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMARMDesc.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMARMAsmPrinter.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMARMInfo.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMARMUtils.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAArch64CodeGen.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMGlobalISel.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAArch64AsmParser.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAArch64Disassembler.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAArch64Desc.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAArch64AsmPrinter.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAArch64Info.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAArch64Utils.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMPowerPCCodeGen.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMPowerPCAsmParser.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMPowerPCDesc.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMPowerPCAsmPrinter.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMPowerPCDisassembler.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMPowerPCInfo.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMSystemZCodeGen.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMSystemZAsmParser.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMSystemZDisassembler.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMSystemZDesc.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMSystemZAsmPrinter.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMSystemZInfo.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMMipsCodeGen.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAsmPrinter.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMDebugInfoCodeView.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMDebugInfoMSF.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMSelectionDAG.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMMipsAsmParser.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMMipsDesc.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMMipsAsmPrinter.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMMipsDisassembler.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMMCDisassembler.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMMipsInfo.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangFrontendTool.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangCodeGen.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMCoroutines.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMCoverage.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMLTO.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMObjCARCOpts.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMPasses.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMipo.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMIRReader.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAsmParser.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMInstrumentation.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMLinker.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMCodeGen.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMBitWriter.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMScalarOpts.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMInstCombine.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMVectorize.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMTransformUtils.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMTarget.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMAnalysis.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMObject.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangARCMigrate.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangStaticAnalyzerFrontend.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangStaticAnalyzerCheckers.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangStaticAnalyzerCore.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangCrossTU.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangIndex.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangFormat.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangToolingCore.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangASTMatchers.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangRewriteFrontend.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangFrontend.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMProfileData.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangDriver.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMOption.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangParse.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMMCParser.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangSerialization.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangSema.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangAnalysis.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangEdit.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangRewrite.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangAST.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangLex.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangAPINotes.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMBitReader.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libclangBasic.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMMC.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMCore.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMBinaryFormat.a lib/libswiftOption.a lib/libswiftBasic.a /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMSupport.a -lz -lrt -ldl -ltinfo -lpthread -lm /home/asb/Developer/swift/build/Ninja-DebugAssert/llvm-linux-x86_64/lib/libLLVMDemangle.a lib/libswiftDemangling.a /usr/lib64/libuuid.so && :
lib/libswiftBasic.a(Statistic.cpp.o): In function `swift::FrontendStatsTracer::FrontendStatsTracer(swift::UnifiedStatsReporter*, llvm::StringRef, swift::Decl const*)':
/home/asb/Developer/swift/swift/lib/Basic/Statistic.cpp:462: undefined reference to `swift::UnifiedStatsReporter::TraceFormatter const* swift::FrontendStatsTracer::getTraceFormatter<swift::Decl const*>()'
lib/libswiftBasic.a(Statistic.cpp.o): In function `swift::FrontendStatsTracer::FrontendStatsTracer(swift::UnifiedStatsReporter*, llvm::StringRef, swift::ProtocolConformance const*)':
/home/asb/Developer/swift/swift/lib/Basic/Statistic.cpp:468: undefined reference to `swift::UnifiedStatsReporter::TraceFormatter const* swift::FrontendStatsTracer::getTraceFormatter<swift::ProtocolConformance const*>()'
lib/libswiftBasic.a(Statistic.cpp.o): In function `swift::FrontendStatsTracer::FrontendStatsTracer(swift::UnifiedStatsReporter*, llvm::StringRef, swift::Expr const*)':
/home/asb/Developer/swift/swift/lib/Basic/Statistic.cpp:472: undefined reference to `swift::UnifiedStatsReporter::TraceFormatter const* swift::FrontendStatsTracer::getTraceFormatter<swift::Expr const*>()'
lib/libswiftBasic.a(Statistic.cpp.o): In function `swift::FrontendStatsTracer::FrontendStatsTracer(swift::UnifiedStatsReporter*, llvm::StringRef, clang::Decl const*)':
/home/asb/Developer/swift/swift/lib/Basic/Statistic.cpp:477: undefined reference to `swift::UnifiedStatsReporter::TraceFormatter const* swift::FrontendStatsTracer::getTraceFormatter<clang::Decl const*>()'
lib/libswiftBasic.a(Statistic.cpp.o): In function `swift::FrontendStatsTracer::FrontendStatsTracer(swift::UnifiedStatsReporter*, llvm::StringRef, swift::SILFunction const*)':
/home/asb/Developer/swift/swift/lib/Basic/Statistic.cpp:482: undefined reference to `swift::UnifiedStatsReporter::TraceFormatter const* swift::FrontendStatsTracer::getTraceFormatter<swift::SILFunction const*>()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

graydon commented:
@compnerd & @AndrewSB If I had to guess, I would think that the not-lld linker I tested with (I guess ld64) is deciding the various `FrontendStatsTracer::FrontendStatsTracer` constructor overloads in Statistic.cpp are dead symbols before it tries to resolve their calls to the forward-declared (and defined in higher-level libraries) `getTraceFormatter<>` specializations, whereas lld is deciding the constructor overloads are not-dead and then trying to resolve them. Or perhaps they wind up as weak references or something. I don't know what's formally correct here but clearly I botched making it portable, and will post a fix as soon as I can reproduce what you're seeing.

(I _suspect_ solving it is therefore just a matter of moving those one-liner constructors from Statistic.cpp to inlines in Statistic.h, so they don't get codegen'ed _at all_ when not in use; but there might be other approaches that work too. If you want to try a short term fix ahead of me, try that?)

compnerd committed:
Basic: attempt to repair inverted dependencies

When linking on Linux, we would fail with unresolved symbol references
to swift::FrontendStatsTracer::getTraceFormatter<T>. The use of the
types, which are defined in swiftSIL occur in swiftBasic.  Provide
inline definitions of the constructors which cause the dependency on
some environments (e.g. Linux).

compnerd commented:
@swift-ci please test

jrose-apple commented on include/swift/Basic/Statistic.h:
```diff
@@ -225,20 +225,61 @@ class FrontendStatsTracer
+// Provide inline definitions for the delegating constructors.  These avoid
+// introducing a circular dependency between libParse and libSIL.  They are
+// marked as `inline` explicitly to prevent ODR violations due to multiple
+// emissions.  We cannot force the inling by defining them in the declaration
```
Typo: "inling".

jrose-apple commented:
If this new patch fixes the issue I see no problems with it.

compnerd commented:
Yeah, it fixes the issue.  I'll update it for the typo.  *sigh* typing is hard.

compnerd commented:
@swift-ci please test and merge

jrose-apple commented:
Please wait for @graydon.

graydon commented:
Looks good to me (though I assumed you'd just put the constructors _in_ the class definition, rather than declaring in the class and defining as inlines immediately outside the class; but I don't especially care). Sorry for the delay in replying!

graydon commented:
Ah never mind, stopped to actually read the comment re "We cannot force the inlining by defining them in the declaration". That's fine then!

jrose-apple commented:
@swift-ci please test

jrose-apple commented:
@swift-ci Please test
